### ISSUE TITLE
parallel helm installs

### ISSUE THREAD

**Issue by skeenan947** [roboll/helmfile#591]
Helmfile should support installing multiple releases at the same time rather than doing them in serial.
In addition, there should be a way to set an order of dependencies, in case one install is dependent on another.

**Comment by mumoshu** [comment]
@skeenan947 Hey!

`helmfile` already support `--concurrency=N` which allows you to set the number of concurrent releases being processed by helmfile, and it's set to the number of releases within the helmfile.yaml by default.

For controlling the installation order, sub-helmfiles listed under `helmfiles:` cna be used, as each sub-helmfile is processed sequentially, whereas the `releases` in a single helmfile is processed concurrently.

Does it help your use-case?

**Comment by skeenan947** [comment]
@mumoshu thank you, I didn't see concurrency in the docs when I looked.  I must've missed it.
re: dependency order, that's helpful, but doesn't completely solve the issue.  I'd love to have more granular control - we can hack something together with multiple Helmfiles, but it'll be clunky and hard to follow.  Would much prefer to have  something like a group# in each release that controls order and defines which Helm charts to install together, and what groups are dependent on previous groups.

**Comment by mumoshu** [comment]
@skeenan947 Hey! Thanks for your response.

> which Helm charts to install together, and what groups are dependent on previous groups.

This looks exactly what `helmfiles:` provides you. Would you mind giving me a configuration example with the desired feature?

**Comment by naseemkullah** [comment]
> `helmfile` already support `--concurrency=N` which allows you to set the number of concurrent releases being processed by helmfile, and it's set to the number of releases within the helmfile.yaml by default.

Hi @mumoshu, does this work out of the box if we are using `helmfile.d/foo.yaml` for example? or does the file need to be called `helmfile.yaml` for `concurrency=#-of-releases` to work?

**Comment by mumoshu** [comment]
@naseemkullah Hey!

Unfortunately, `helmfile.d` is designed as it sort all the files under the directory in alphabetical order, and runs it sequentially. In each helmfile all the releases are processed concurrently, which is configurable via `--concurrency`.

So yes, it must be a single `helmfile.yaml` (although the name doesn't matter) to make the releases processed concurrently. This is unlikely to change, as a helmfile state file is treated as an unit of job, so that `helmfile.d` or sub-helmfiles can be used for ordering.

But if your helmfiles are simple enough, you should be able to `bases:` to merge all the helmfile state files under your `helmfile.d`, which produces a single big helmfile state file that all the releases originally defined in different state files are processed concurrently.

**Comment by naseemkullah** [comment]
Thanks @mumoshu!

**Comment by abdennour** [comment]
Does `--concurrency=N` is still supported with the latest versions ?

**Comment by mumoshu** [comment]
Yes

**Comment by razsha** [comment]
Hey @mumoshu !
I understand the idea of treating a helmfile as a single unit. but, you might want the ability to both run a single "component" (sub-helmfile) + run all components in parallel. even more, you might want to run one 'tier' in parallel and another serially. 
I thought about 2 options:
1. A helmfile flag that disables this behavior of sequentially.
2. Add to `helmfiles:` a sub attribute for folders you want to run in parallel.

WDYT?

**Comment by mumoshu** [comment]
@razsha Hey! I'm not saying no to the request feature but I do have concerns.

The first option breaks declarativeness of helmfile.yaml, cuz it alters the meaning of a helmfile.yaml depending on a command-line flag.

So the latter would be a better direction.

But note that currently any `helmfiles:` entry is processed sequentially assuming the order matters, and files in a directory specified via `helmfiles:` is also processed sequentially in alphabetical order assuming the user is using file names to control the order. So depending on the use-case, we may need to make both configurable. But what would be the best API to express that?

Also, helmfile has an alterative mode that allows you to specify a directory containing multiple helmfile.yaml files via `helmfile -d PATH/TO/DIR`. If you make the file ordering configurable, we also need to make it configurable via a command-line flag, and it breaks the declarativeness of the dir cuz the installation order within a directory is now dependent on the flag.

Given all that, I guess our best be would be to give up controlling order of files under the directory, and only allow controlling the order and the parallelism of `helmfiles:` entry, extending the DAG feature(`releases[].needs`) to sub-helmfiles.

```
helmfiles:
- name: infra
  path: helmfile.infra.yam
- name: apps1
  path: helmfile.apps1.yaml 
  needs: infra
- name: apps2
  path: helmfile.apps2.yaml 
  needs: infra
```

This results Helmfile computing a DAG of helmfiles `[ infra ] <-- depends on [app1, app2]`. As app1 and app2 being independent to each other after `infra` is given, helmfile will automatically install `infra` first, and then concurrently install `app1` and `app2`.

Would that work for you?

**Comment by razsha** [comment]
Thanks @mumoshu !
Yes, I think it's a good idea to express it via `helmfiles:`. it might be a breaking change for everyone who assumes that sub-helmfiles run sequentially, but the `needs` is a good alternative.

**Comment by ostlerc** [comment]
am I understanding this correctly?
if you have a globbing top level helmfile like 
```yaml
helmfiles:
- apps/*/helmfile.yaml
```
then all nested helmfiles would run in serial?

**Comment by mumoshu** [comment]
Yes

**Comment by ostlerc** [comment]
I was looking into doing this work, but don't want to make a large change such as the DAG right now. I was toying around with the idea of adding another flag for parallel files. this flag would represent how many helmfiles to process in parallel. I think this would be a good step towards solving this issue. Longer term there could be improvments around the helmstate  DAG . My current use case is quite simple though, we just want our deploys to go faster with a globbing organization of files.

While digging through I also realized that the sync and apply code have a large portion of their code copy pasted with small changes. I don't know enough of the differences between the two in detail. My understanding from the documentation is that apply simply does a diff then a sync. so logically I thought the apply would reuse the sync code after a diff code was executed. Any light you can shed on this?

I am actively looking to add these changes and feedback about these ideas would be appreciated so I don't go down the wrong path.

**Comment by sergusha** [comment]
Has --concurrency been removed from helmfile? I'm ending up with an error
`flag provided but not defined: -concurrency`

**Comment by FACOLOMBANI** [comment]
Hi! Any news about this topic, i'm encountering same issue while trying to deploy 15 components and need to wait for about 30 mins for everything to run because of the serial deployment.

**Comment by ostlerc** [comment]
We ended up running helmfile multiple times in parallel.

**Comment by mumoshu** [comment]
Feel free to submit a pull request to address this. Note also that we've already moved to `helmfile/helmfile`!
